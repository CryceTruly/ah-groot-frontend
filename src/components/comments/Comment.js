import React from 'react';
import CommentCreateForm from './CommentForm';

const Comment = props => {
  return (
    <div className="comment-container">
      <div className="comment-owner">{props.user}</div>
      <div className="comment-timestamp">{props.createdAt}</div>
      <span>
        <i
          className="small material-icons right comment-update"
          onClick={props.deleteComment}
        >
          delete
        </i>
      </span>
      <span>
        <i
          className="small material-icons right comment-update"
          onClick={props.toggleEditForm}
        >
          edit
        </i>
      </span>
      <div className="comment-body card-panel">
        <div className="comment-body">{props.body}</div>
      </div>
    </div>
  );
};

export default Comment;
