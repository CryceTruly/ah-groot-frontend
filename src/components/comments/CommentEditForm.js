import React from 'react';
import '../../styles/comments.scss';

const CommentEditForm = props => {
  return (
    <div className="comment-form">
      <div className="row">
        <form className="col s12">
          <div className="row">
            <div className="input-field col s12">
              <textarea
                id="textarea1"
                className="materialize-textarea"
                value={props.commentText}
                onChange={props.commentTextChange}
              />
              <label htmlFor="textarea1" className="comment-label">
                Edit your comment here
              </label>
            </div>
          </div>
          <input
            type="submit"
            value="Save"
            className="btn right"
            onClick={props.handleSubmit}
          />
          <button className="btn right" onClick={props.toggleCommentEditForm}>
            Cancel
          </button>
        </form>
      </div>
    </div>
  );
};

export default CommentEditForm;
