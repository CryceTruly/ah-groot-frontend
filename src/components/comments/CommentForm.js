import React from 'react';
import '../../styles/comments.scss';

const CommentForm = props => {
  return (
    <div className="comment-create-form">
      <div className="row">
        <form className="col s12">
          <div className="row">
            <div className="input-field col s12">
              <textarea
                id="textarea1"
                className="materialize-textarea"
                value={props.commentText}
                onChange={props.commentTextChange}
              />
              <label htmlFor="textarea1" className="comment-label">
                {props.textareaLabel}
              </label>
            </div>
          </div>
          <input
            type="submit"
            value={props.inputValue}
            className="btn right"
            onClick={props.handleSubmit}
          />
        </form>
      </div>
    </div>
  );
};

export default CommentForm;
