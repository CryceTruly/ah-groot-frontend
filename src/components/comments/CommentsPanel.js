import React from 'react';
//
import Comment from '../../containers/Comments/Comment';

const CommentsPanel = props => {
  const comments = props.comments ? props.comments : [];
  const commentItems = comments.map(comment => (
    <Comment
      key={comment.id}
      id={comment.id}
      body={comment.body}
      user={comment.user.user}
      createdAt={comment.created_at}
      slug={props.slug}
      getComments={props.getComments}
    />
  ));

  return (
    <div>
      <section className="card">
      <p className="comments-title center-align">Comments</p>
      <hr className="comment-title-divider"></hr>
        <ul>{commentItems}</ul>
      </section>
    </div>
  );
};

export default CommentsPanel;
