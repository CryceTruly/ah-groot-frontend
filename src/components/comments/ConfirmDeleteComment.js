import React from 'react';
import '../../styles/comments.scss';

const ConfirmDeleteComment = props => {
  return (
    <div className="comment-delete center-align">
      <p className="comment-delete">Are you sure you want to delete this comment?</p>
      <button className="btn cancel" onClick={props.toggleConfirmDelete}>
        Cancel
      </button>
      <button className="btn delete" onClick={props.deleteComment}>
        Delete
      </button>
    </div>
  );
};

export default ConfirmDeleteComment;
