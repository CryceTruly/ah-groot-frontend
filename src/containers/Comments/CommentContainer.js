import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getComments, postComment } from '../../actions/commentAction';
import CommentForm from '../../components/comments/CommentForm';
import CommentsPanel from '../../components/comments/CommentsPanel';

export class CommentContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      displayCommentForm: false,
      commentText: '',
    };
  }

  componentWillMount() {
    this.props.getComments(this.props.slug);
  }

  toggleCommentForm = () => {
    this.setState({
      displayCommentForm: !this.state.displayCommentForm,
    });
  };

  commentTextChange = event =>
    this.setState({ commentText: event.target.value });

  createComment = event => {
    event.preventDefault();
    const commentText = this.state.commentText;
    this.props.postComment(commentText, this.props.slug);
    this.setState({ displayCommentForm: false });
    document.location.reload(true);
  };

  render() {
    return (
      <div className="comments">
        <div>
          <CommentsPanel
            comments={this.props.comments}
            slug={this.props.slug}
            getComments={this.props.getComments}
          />
        </div>

        <button className="btn comment-add" onClick={this.toggleCommentForm}>
          Add Comment
        </button>

        {this.state.displayCommentForm ? (
          <div>
            <CommentForm
              handleSubmit={this.createComment}
              commentText={this.state.commentText}
              commentTextChange={this.commentTextChange}
              inputValue="Post Comment"
              textareaLabel="Write your comment here"
            />
          </div>
        ) : (
          <div />
        )}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  comments: state.comments.comments,
});

const mapDispatchToProps = {
  getComments,
  postComment,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CommentContainer);
