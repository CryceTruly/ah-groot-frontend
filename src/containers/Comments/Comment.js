import React, { Component } from 'react';
import {
  updateComment,
  deleteComment,
  likeComment,
} from '../../actions/commentAction';
import CommentEditForm from '../../components/comments/CommentEditForm';
import ConfirmDeleteComment from '../../components/comments/ConfirmDeleteComment';
import { connect } from 'react-redux';
import moment from 'moment';

export class Comment extends Component {
  constructor(props) {
    super(props);
    this.state = {
      displayEditCommentForm: false,
      displayConfirmDelete: false,
      commentText: '',
    };
  }

  toggleEditForm = () => {
    this.setState({
      displayEditCommentForm: !this.state.displayEditCommentForm,
    });
  };

  toggleConfirmDelete = () => {
    this.setState({
      displayConfirmDelete: !this.state.displayConfirmDelete,
    });
  };

  commentTextChange = event =>
    this.setState({ commentText: event.target.value });

  updateComment = event => {
    event.preventDefault();
    this.setState({ displayEditCommentForm: false });
    const commentText = this.state.commentText;
    this.props.updateComment(commentText, this.props.id, this.props.slug);
  };

  getDateCreated = () => {
    return `${moment(`${new Date(this.props.createdAt)}`).fromNow()}`;
  };

  deleteComment = () => {
    this.setState({ displayConfirmDelete: false });
    this.props.deleteComment(this.props.id, this.props.slug);
  };

  likeComment = () => {
    this.props.likeComment(this.props.id, this.props.slug);
  };

  render() {
    return (
      <div className="comment-container">
        <div className="comment-owner">
          {this.props.user} <span className="commented">commented: </span>
        </div>
        <div className="comment-timestamp">{this.getDateCreated()}</div>
        {this.props.user === sessionStorage.user ? (
          <div>
            <span>
              <i
                className="small material-icons right comment-edit one"
                onClick={this.toggleConfirmDelete}
              >
                delete
              </i>
            </span>
            <span>
              <i
                className="small material-icons right comment-edit two"
                onClick={this.toggleEditForm}
              >
                edit
              </i>
            </span>
          </div>
        ) : (
          <div />
        )}

        <span className="right">
          <i
            className="small material-icons comment-edit three"
            onClick={this.likeComment}
          >
            thumb_up
          </i>
          <span className="comment-like-count">{this.props.likes}</span>
        </span>
        <div className="comment-body card-panel">
          <div className="comment-body">{this.props.body}</div>
        </div>

        {this.state.displayEditCommentForm ? (
          <div>
            <CommentEditForm
              commentText={this.state.commentText}
              commentTextChange={this.commentTextChange}
              handleSubmit={this.updateComment}
              toggleEditForm={this.toggleEditForm}
            />
          </div>
        ) : (
          <div />
        )}

        {this.state.displayConfirmDelete ? (
          <div>
            <ConfirmDeleteComment
              deleteComment={this.deleteComment}
              toggleConfirmDelete={this.toggleConfirmDelete}
            />
          </div>
        ) : (
          <div />
        )}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  commentId: state.commentId,
});

const mapDispatchToProps = {
  updateComment,
  deleteComment,
  likeComment,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Comment);
