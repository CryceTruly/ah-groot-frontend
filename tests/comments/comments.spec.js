import configureStore from 'redux-mock-store';
import React from 'react';
import { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import CommentContainer from '../../src/containers/Comments/CommentContainer';
import Comment from '../../src/containers/Comments/Comment';
import commentData from './commentData';
import CommentForm from '../../src/components/comments/CommentForm';
import CommentEditForm from '../../src/components/comments/CommentEditForm';
import ConfirmDeleteComment from '../../src/components/comments/ConfirmDeleteComment';
import CommentsPanel from '../../src/components/comments/CommentsPanel';

describe('comment container', () => {
  const initialState = {
    commentPosted: false,
    commentUpdated: false,
    commentDeleted: false,
    comments: [],
  };
  const mockStore = configureStore([thunk]);
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it('should render without crashing', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <CommentContainer slug={'an-article'} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a comment form', () => {
    const wrapper = mount(
      <Provider store={store}>
        <CommentForm />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a comment edit form', () => {
    const wrapper = mount(
      <Provider store={store}>
        <CommentEditForm />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a confirm delete comment component', () => {
    const wrapper = mount(
      <Provider store={store}>
        <ConfirmDeleteComment />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a comments panel', () => {
    const wrapper = mount(
      <Provider store={store}>
        <CommentsPanel comments={[commentData]} />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the comment container with its children', () => {
    const wrapper = mount(
      <Provider store={store}>
        <CommentContainer slug={'an-article'} />
      </Provider>
    );
    wrapper.setState({ displayCommentForm: false, commentText: '' });
    const event = {
      target: { value: '' },
    };
    wrapper.find('button.btn').simulate('click');
    wrapper
      .find('textarea.materialize-textarea')
      .simulate('change', { target: { value: 'l' } });
    wrapper.find('input.btn.right').simulate('click');
  });

  it('should render a single comment', () => {
    const displayAllComments = jest.fn();
    const wrapper = mount(
      <Provider store={store}>
        <Comment />
      </Provider>
    );
    wrapper.setProps({
      key: commentData.id.toString,
      id: commentData,
      body: commentData.body,
      user: commentData.user.user,
      createdAt: commentData.created_at.toString(),
      slug: 'an-article',
      displayAllComments: displayAllComments,
    });

    wrapper.setState({
      displayEditCommentForm: true,
      displayConfirmDelete: true,
      commentText: 'comment',
    });

    wrapper
      .find('i.small.material-icons.right.comment-edit.one')
      .simulate('click');
    wrapper
      .find('i.small.material-icons.right.comment-edit.two')
      .simulate('click');
    wrapper.find('i.small.material-icons.comment-edit.three').simulate('click');
    wrapper
      .find('textarea.materialize-textarea.comment-edit')
      .simulate('change', { target: { value: 'l' } });
    wrapper.find('input.btn.right').simulate('click');
    wrapper.find('button.btn.delete').simulate('click');
  });
});
