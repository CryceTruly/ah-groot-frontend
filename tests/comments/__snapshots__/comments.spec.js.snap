// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comment container should render a comment edit form 1`] = `ReactWrapper {}`;

exports[`comment container should render a comment form 1`] = `ReactWrapper {}`;

exports[`comment container should render a comments panel 1`] = `ReactWrapper {}`;

exports[`comment container should render a confirm delete comment component 1`] = `ReactWrapper {}`;

exports[`comment container should render without crashing 1`] = `ShallowWrapper {}`;
